src/tmpfiles/tmpfiles.c

```diff
@@ -2004,6 +2006,64 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
         return r;
 }
 
+static int rm_if_wrong_type(mode_t mode, const char *path, bool follow) {
+        struct stat st;
+        int r;
+
+        assert(path);
+        assert((mode & ~S_IFMT) == 0);
+        log_debug("Checking type of %s", path);
+
+        if (follow)
+                r = stat(path, &st);
+        else
+                r = lstat(path, &st);
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        log_error_errno(r, "stat(%s): %m", path);
+                }
+                return -errno;
+        }
+
+        if ((st.st_mode & S_IFMT) == mode)
+                return 0;
+
+        log_warning("wrong file type 0x%x; rm -rf \"%s\"", st.st_mode & S_IFMT, path);
+        r = rm_rf(path, REMOVE_ROOT|REMOVE_SUBVOLUME|REMOVE_PHYSICAL);
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        log_error_errno(r, "rm_rf(%s): %m", path);
+                }
+                return -errno;
+        }
+
+        return 0;
+}
+
+static int rm_nondir_parents(const char *path) {
+        char *p, *e;
+        int r = -ENOENT;
+
+        assert(path);
+
+        // Walk up the path components until one is found with the expected type or there are no more.
+        p = strdupa(path);
+        // If the path doesn't exist, check the next path component.
+        while (r == -ENOENT || r == -ENOTDIR) {
+                e = strrchr(p, '/');
+                if (!e)
+                        return 0;
+
+                *e = 0;
+                r = rm_if_wrong_type(S_IFDIR, p, true);
+                // Remove dangling symlinks.
+                if (r == -ENOENT) {
+                        r = rm_if_wrong_type(S_IFDIR, p, false);
+                }
+        }
+        return r;
+}
```

Comment:
so this will go wrong, in case you have /foo/bar/baz and /foo is already a symlink. because in that case you want to remove /foo and are done, and leave /foo/bar and /foo/bar/baz alone.

i.e. you need to work front to back, not back to front.

Moreover, for new code i'd really prefer if we'd go through this with openat(), i.e. start at the root dir, then use opanat() relative to it to open the first subdir with O_DIRECTORY. If that fails with ENOTDIR, issue unlinkat() instead and then issue mkdirat() to create the new dir. if you see ENOENT immediately create the new dir. on all other errors fail. After creating each subdir open the subdir immediately with O_DIRECTORY, and work on the next element, and so on.

Reply:
In your example if /foo was a symlink pointed to a directory, it would not be removed by this function. If it was a broken symlink or didn't point to a directory it would get removed, but then /foo/bar wouldn't exist. This is moot though if we go with openat which is fine with me.

I didn't use openat because, for it to really provide protection against time-of-check-time-of-use attacks, the code would need to be more integrated (i.e. part of mkdir_parents_label). I fine with uploading a rewrite that covers mkdirat for the parents as well, but that doesn't address the issues with the rest of the code still operating on path strings instead of reusing fds opened with O_DIRECTORY that were already checked. I am curious if there is interest in pursuing improvements along these lines in follow up changes.

Reply:
my suggestion would be to implement this as alternative to mkdir_parents_label(), instead of in addition. i.e. in the new "=" mode, use your new function, and without it use mkdir_parents_label() as before, but never both.

In general for all new code we should go the openat() way as much as we can, to make things safer.

Reply:
maybe add a helper mkdir_parents_harder() or so that is like mkdir_parents_label(), but does this replace-everything-by-directory dance

Reply:
I wrote the helper function.